ledger/participant-integration-api/src/app/scala/com/daml/ledger/metering/Main.scala

```diff
@@ -0,0 +1,128 @@
+import com.daml.platform.apiserver.meteringreport.{JcsSigner, MeteringReportKey}
+
+import java.nio.charset.StandardCharsets
+import java.nio.file.{Files, Path}
+import scala.jdk.StreamConverters._
+import scala.util.Try
+
+object Main {
+
+  trait ExitCode {
+    def code: String
+    def detail: String
+    final def message = s"$code: $detail"
+  }
+
+  type ExitCodeOr[T] = Either[ExitCode, T]
+
+  abstract class Code(val code: String) extends ExitCode
+  case class CodeDetail(code: String, detail: String) extends ExitCode
+
+  val OK: ExitCode = CodeDetail("OK", "The digest is as expected")
+
+  case class ErrUsage(app: String) extends Code("ERR_USAGE") {
+    def detail = s"Usage: $app <directory> <report>"
+  }
+
+  case class NotDirectory(dir: String) extends Code("ERR_NOT_DIRECTORY") {
+    def detail = s"The passed directory is not a valid directory: $dir"
+  }
+
+  case class NotKeyFile(path: Path, t: Throwable) extends Code("ERR_NOT_KEY_FILE") {
+    def detail = s"Unable to parse metering report key from key file: $path [${t.getMessage}]"
+  }
+
+  case class NoKeys(dir: String) extends Code("ERR_NO_KEYS") {
+    def detail = s"No keys found in the key directory: $dir"
+  }
+
+  case class NotFile(report: String) extends Code("ERR_NO_REPORT") {
+    def detail = s"The passed report is not a file: $report"
+  }
+
+  case class FailedToReadReport(report: String, t: Throwable) extends Code("ERR_READING_REPORT") {
+    def detail = s"Failed to read the participant report: $report [${t.getMessage}]"
+  }
+
+  case class FailedVerification(status: VerificationStatus)
+      extends Code("ERR_FAILED_VERIFICATION") {
+    def detail = s"Report verification failed with status: $status"
+  }
+
+  def main(args: Array[String]): Unit = {
+    val result = for {
+      keyReport <- checkUsage(args)
+      (keyDir, reportFile) = keyReport
+      report <- readReport(reportFile)
+      keys <- readKeys(keyDir)
+      _ <- verifyReport(report, keys)
+    } yield {
+      ()
+    }
+    result match {
+      case Right(_) =>
+        System.out.println(OK.message)
+        System.exit(0)
+      case Left(exitCode) =>
+        System.err.println(exitCode.message)
+        System.exit(1)
+    }
+  }
+
+  private def checkUsage(args: Array[String]): ExitCodeOr[(String, String)] = {
+    args.toList match {
+      case List(dir, report) => Right((dir, report))
+      case _ => Left(ErrUsage("metering-verification-app"))
+    }
+  }
+
+  private def readKey(keyPath: Path): ExitCodeOr[Key] = {
+    Try(MeteringReportKey.assertParseKey(Files.readAllBytes(keyPath)))
+      .fold(t => Left(NotKeyFile(keyPath, t)), Right(_))
+  }
+
+  private def readKeys(keyDir: String): ExitCodeOr[Map[String, Key]] = {
+
+    import scalaz._
+    import scalaz.syntax.traverse._
+    import std.either._
+    import std.list._
+
+    val dir = Path.of(keyDir)
+    if (Files.isDirectory(dir)) {
+      for {
+        keys <- Files.list(dir).toScala(List).traverse(readKey)
```

Comment:
Wouldn't it be better to allow the folder to contain both key files as well as other files?

Reply:
Only allowing key files is simplest:
- Directories are cheap
- We don't need to filter out files based on some file name pattern criteria.
- We can fail with an error if any file in the directory is not a valid key file.